pkg/target/http.go

```diff
@@ -226,52 +268,175 @@ func (ht *HTTPTarget) Write(messages []*models.Message) (*models.TargetWriteResu
-			failed = append(failed, msg)
-			continue
+			ht.monitoring.SendAlert(monitoring.Alert{Message: fmt.Sprintf("Could not create HTTP request, error - %s", err.Error())})
+			health.SetUnhealthy()
+			return nil, err
 		}
-		request.Header.Add("Content-Type", ht.contentType)                // Add content type
-		addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(msg)) // Add headers if there are any
-		if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {     // Add basic auth if set
+
+		request.Header.Add("Content-Type", ht.contentType)
+		addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(msg))
+		if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {
 			request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
 		}
+		return nil, err
+	})
+
+	return result.(*http.Request)
+}
+
+func (ht *HTTPTarget) executeHTTPRequest(request *http.Request, msg *models.Message) Response {
+	// Second possible place for failure - error after making HTTP call. We don't have HTTP response, so there is no way to check status/response body.
+	// It's connection/response timeout or some other unexpected network issue.
+	// Could we categorize it as transient error? So it deserves bunch of retries, after reaching max attempt just fail.
+	transientConfig := ht.responseHandler.RetryStrategies["transient"]
+	result := ht.retry(transientConfig, func() (interface{}, error) {
 		requestStarted := time.Now()
-		resp, err := ht.client.Do(request) // Make request
+		resp, err := ht.client.Do(request)
 		requestFinished := time.Now()
 
 		msg.TimeRequestStarted = requestStarted
 		msg.TimeRequestFinished = requestFinished
-
 		if err != nil {
-			errResult = multierror.Append(errResult, err)
-			failed = append(failed, msg)
-			continue
+			health.SetUnhealthy()
+			return nil, err
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		return &Response{Body: string(body), Status: resp.StatusCode}, nil
+	})
+
+	return result.(Response)
+}
+
+// We have HTTP response so we can check status and response body details.
+// Returned error means we have to retry
+func (ht *HTTPTarget) handleHTTPResponse(request *http.Request, response Response, msg *models.Message, sent []*models.Message, invalid []*models.Message, failed []*models.Message, errResult error, previouslyMatchedRule *Rule) error {
+	//ACK and set healthy on success...
+	if ht.isSuccess(response) {
+		health.SetHealthy()
+		sent = append(sent, msg)
+		if msg.AckFunc != nil {
+			msg.AckFunc()
 		}
-		defer resp.Body.Close()
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-			sent = append(sent, msg)
-			if msg.AckFunc != nil { // Ack successful messages
-				msg.AckFunc()
-			}
-		} else {
-			errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-			failed = append(failed, msg)
-			continue
+		return nil
+	}
+
+	if ht.isInvalid(response) {
```

Comment:
Oh this is very interesting. Configuring the flow of data around the app as well. Nice idea